evm/src/cpu/syscalls.rs

```diff
@@ -2,62 +2,81 @@
 ) {
-    let syscall_list = Lazy::force(&TRAP_LIST);
-    // 1 if _any_ syscall, else 0.
-    let should_syscall: P = syscall_list
-        .iter()
-        .map(|&(col_index, _)| lv[col_index])
+    let filter = lv.is_cpu_cycle * lv.op.syscall;
+
+    // Look up the handler in memory
+    let code_segment = P::Scalar::from_canonical_usize(Segment::Code as usize);
+    let syscall_jumptable_start =
+        P::Scalar::from_canonical_usize(KERNEL.global_labels["syscall_jumptable"]);
+    let opcode: P = lv
+        .opcode_bits
+        .into_iter()
+        .enumerate()
+        .map(|(i, bit)| bit * P::Scalar::from_canonical_u64(1 << i))
         .sum();
-    let filter = lv.is_cpu_cycle * should_syscall;
+    let opcode_handler_addr_start =
+        syscall_jumptable_start + opcode * P::Scalar::from_canonical_usize(BYTES_PER_OFFSET);
+    for (i, channel) in lv.mem_channels[0..BYTES_PER_OFFSET].iter().enumerate() {
+        yield_constr.constraint(filter * (channel.used - P::ONES));
+        yield_constr.constraint(filter * (channel.is_read - P::ONES));
+
+        // Set kernel context and code segment
+        yield_constr.constraint(filter * channel.addr_context);
+        yield_constr.constraint(filter * (channel.addr_segment - code_segment));
 
-    // If syscall: set program counter to the handler address
-    // Note that at most one of the `lv[col_index]`s will be 1 and all others will be 0.
-    let syscall_dst: P = syscall_list
+        // Set address, using a separate channel for each of the `BYTES_PER_OFFSET` limbs.
+        let limb_address = opcode_handler_addr_start + P::Scalar::from_canonical_usize(i);
+        yield_constr.constraint(filter * (channel.addr_virtual - limb_address));
+    }
+
+    // Disable unused channels (the last channel is used to push to the stack)
+    for channel in &lv.mem_channels[BYTES_PER_OFFSET..NUM_GP_CHANNELS - 1] {
+        yield_constr.constraint(filter * channel.used);
+    }
+
+    // Set program counter to the handler address
+    // The addresses are big-endian in memory
+    let target = lv.mem_channels[0..BYTES_PER_OFFSET]
         .iter()
-        .map(|&(col_index, handler_addr)| {
-            lv[col_index] * P::Scalar::from_canonical_usize(handler_addr)
-        })
-        .sum();
-    yield_constr.constraint_transition(filter * (nv.program_counter - syscall_dst));
-    // If syscall: set kernel mode
+        .map(|channel| channel.value[0])
+        .fold(P::ZEROS, |cumul, limb| {
+            cumul * P::Scalar::from_canonical_u64(256) + limb
+        });
+    yield_constr.constraint_transition(filter * (nv.program_counter - target));
+    // Set kernel mode
     yield_constr.constraint_transition(filter * (nv.is_kernel_mode - P::ONES));
+    // Maintain current context
+    yield_constr.constraint_transition(filter * (nv.context - lv.context));
 
+    // This memory channel is constrained in `stack.rs`.
     let output = lv.mem_channels[0].value;
```

Comment:
Should the push be changed to be on the last channel now?

Reply:
Yes, it should! Great catch!